### ISSUE TITLE
set_global on an immutable validates, as do get_global/get_local of missing ones

### ISSUE THREAD

**Issue by dcodeIO** [WebAssembly/binaryen#1317]
Just stumbled upon this. As of [Semantics](http://webassembly.org/docs/semantics/#global-variables)

> It is a validation error for a set_global to index an immutable global variable.

```js
var mod = new binaryen.Module();
mod.addGlobal("a", binaryen.i32, false, mod.i32.const(0));

var funcType = mod.addFunctionType("v", binaryen.none, []);
var func = mod.addFunction("start", funcType, [], mod.block("", [
  mod.setGlobal("a", mod.i32.const(1))
]));
mod.setStart(func);

console.log(mod.emitText());
if (mod.validate())
  console.log("-> validates");
```

```s
(module
 (type $v (func))
 (global $a i32 (i32.const 0))
 (memory $0 0)
 (start $start)
 (func $start (; 0 ;) (type $v)
  (set_global $a
   (i32.const 1)
  )
 )
)

-> validates
```

**Comment by kripken** [comment]
We should check if there is a spec test for this. If there isn't, we should add up upstream. If there is and we are missing it because our spec tests are too old, we may need to update them, or add testing for this in another way.

**Comment by binji** [comment]
> We should check if there is a spec test for this...

There's this one: https://github.com/WebAssembly/spec/blob/master/test/core/globals.wast#L50

**Comment by dcodeIO** [comment]
Related: `get_global` on a missing global validates, but ultimately errors with an assertion error:

```js
var mod = new binaryen.Module();
var funcType = mod.addFunctionType("v", binaryen.void, []);
var func = mod.addFunction("test", funcType, [],
  mod.block("", [
    mod.drop(
      mod.getGlobal("missing", binaryen.i32)
    )
  ])
);
mod.addExport("test", func);

if (mod.validate())
  console.log("-> validates");
mod.emitBinary(); // -> Assertion failed: mappedGlobals.count(name), at: binaryen/src/wasm/wasm-binary.cpp,355,getGlobalIndex at Error
```

**Comment by dcodeIO** [comment]
Related: `get_local` on a missing local validates:

```js
var mod = new binaryen.Module();
var funcType = mod.addFunctionType("v", binaryen.void, []);
var func = mod.addFunction("test", funcType, [],
  mod.block("", [
    mod.drop(
      mod.getLocal(0, binaryen.i32)
    )
  ])
);
mod.addFunctionExport("test", "test", func);

console.log(mod.emitText());
if (mod.validate()) {
  console.log("-> validates");
  var binary = mod.emitBinary();
  new WebAssembly.Module(binary); // CompileError: WebAssembly.Module(): Compiling wasm function #0:test failed: invalid local index: 0 @+34
}
```

```
(module
 (type $v (func))
 (memory $0 0)
 (export "test" (func $test))
 (func $test (; 0 ;) (type $v)
  (drop
   (get_local $0)
  )
 )
)

-> validates
CompileError: WebAssembly.Module(): Compiling wasm function #0:test failed: invalid local index: 0 @+34
```

**Comment by kripken** [comment]
Thanks for the testcases, fix for those two is in #1347
